contracts/libraries/LibArray.sol

```diff
@@ -0,0 +1,311 @@
+  ) internal pure returns (uint256[] memory filteredKeys, uint256 nFiltered) {
+    unchecked {
+      uint256 length = keys.length;
+      filteredKeys = new uint256[](length);
+
+      for (uint256 i; i < length; ++i) {
+        if (filterFn(i, values)) {
+          filteredKeys[nFiltered++] = keys[i];
+        }
+      }
+
+      unsafeResize(filteredKeys, nFiltered);
+
+      return (filteredKeys, nFiltered);
+    }
+  }
+
+  /**
+   * @dev Sorts array of addresses `self` based on `values`.
+   * WARNING This function DOES modifies the original `self` and `values`.
+   */
+  function inplaceSortByValue(
+    address[] memory self,
+    uint256[] memory values
+  ) internal pure returns (address[] memory sorted) {
+    return toAddresses(inplaceQuickSortByValue(toUint256s(self), values));
+  }
+
+  /**
+   * @dev Resize a memory array.
+   *
+   * WARNING: The new length of the array should not be greater than the current length to avoid collision with other already allocated memory.
+   */
+  function unsafeResize(uint256[] memory self, uint256 length) internal pure returns (uint256[] memory resized) {
+    assembly ("memory-safe") {
+      resized := self
+      mstore(resized, length)
+    }
+  }
+
+  /**
+   * @dev Resize a memory address array.
+   *
+   * WARNING: The new length of the array should not be greater than the current length to avoid collision with other already allocated memory.
+   */
+  function unsafeResize(address[] memory self, uint256 length) internal pure returns (address[] memory resized) {
+    return toAddresses(unsafeResize(toUint256s(self), length));
+  }
+
+  /**
+   * @dev Sorts `self` based on `values`.
+   * WARNING This function DOES modifies the original `self` and `values`.
+   */
+  function inplaceSortByValue(
+    uint256[] memory self,
+    uint256[] memory values
+  ) internal pure returns (uint256[] memory sorted) {
+    return inplaceQuickSortByValue(self, values);
+  }
+
+  /**
+   * @dev Quick sort `self` based on `values`.
+   * WARNING This function modify `self` and `values`
+   */
+  function inplaceQuickSortByValue(
+    uint256[] memory self,
+    uint256[] memory values
+  ) internal pure returns (uint256[] memory sorted) {
+    uint256 length = self.length;
+    if (length != values.length) revert ErrLengthMismatch();
+    unchecked {
+      if (length > 1) inplaceQuickSortByValue(self, values, 0, int256(length - 1));
+    }
+
+    assembly ("memory-safe") {
+      sorted := self
+    }
+  }
+
+  /**
+   * @dev Internal function to perform quicksort on an `values` based on a corresponding `arr`.
+   * WARNING This function modify `arr` and `values`
+   */
+  function inplaceQuickSortByValue(
+    uint256[] memory arr,
+    uint256[] memory values,
+    int256 left,
+    int256 right
+  ) private pure {
+    unchecked {
+      if (left == right) return;
+      int256 i = left;
+      int256 j = right;
+      uint256 pivot = values[uint256(left + right) >> 1];
```

Comment:
The `left + right` may cause to overflow/underflow issue, right?
```suggestion
      uint256 pivot = values[uint256(left + right) >> 1];
```

Reply:
https://stackoverflow.com/questions/3816446/how-can-i-safely-average-two-unsigned-ints-in-c

Reply:
TIl